runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslator.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction;
+
+import java.util.List;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.coders.Coder;
+
+/**
+ * An interface that translates coders to components and back.
+ *
+ * <p>This interface is highly experimental, and incomplete. Coders must in the general case have
+ * the capability to encode an additional payload, which is not currently supported. This exists as
+ * a temporary measure.
+ */
+@Experimental(Kind.CORE_RUNNERS_ONLY)
+public interface CoderTranslator<T extends Coder<?>> {
```

Comment:
Yes, the `Any` was meant as way of ensuring that the encoded bytes really do represent what the user is expecting in addition to the potential polymorphic aspect, and also as a way to tell people to use proto messages to define their coder schema so fields have descriptions, type safety, auto-generated code, etc...

I originally had a proposal which used a `map<string, bytes>` on the FunctionSpec a long while back but that was too loosely typed and several individuals suggested that we urge people to use an actual proto message and it seemed at the time that using an `Any` best leveraged the point that we want to stick in a proto message and not to use some loose type. So from the FunctionSpec definition perspective, this made the most sense.

For the Java interface, the argument of bytes vs message has a different perspective, but I still believe that it should be either a message or an encoded message (either already wrapped as an `Any` or as raw bytes).